unsafe-fields/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+//!         #[unsafe]
+//!         lo: usize,
+//!         #[unsafe]
+//!         hi: usize,
+//!     }
+//! }
+//!
+//! impl Range {
+//!     pub fn swap(&mut self) {
+//!         // ERROR: Mismatched types
+//!         core::mem::swap(&mut self.lo, &mut self.hi);
+//!     }
+//! }
+//! ```
+//!
+//! # Limitations
+//!
+//! Note that we cannot prevent `Unsafe`s from being swapped between the same
+//! field in instances of the same type:
+//!
+//! ```
+//! use unsafe_fields::{unsafe_fields, Unsafe};
+//!
+//! unsafe_fields! {
+//!     /// A `usize` which is guaranteed to be even.
+//!     pub struct EvenUsize {
+//!         // INVARIANT: `n` is even.
+//!         #[unsafe]
+//!         n: usize,
+//!     }
+//! }
+//!
+//! pub fn swap(a: &mut EvenUsize, b: &mut EvenUsize) {
+//!     core::mem::swap(&mut a.n, &mut b.n);
+//! }
+//! ```
+
+use core::marker::PhantomData;
+
+/// A field with safety invariants.
+///
+/// `Unsafe` should not be named directly - instead, use [`unsafe_fields!`] to
+/// declare a type with unsafe fields.
+///
+/// See the [crate-level documentation](crate) for more information.
+#[repr(transparent)]
+pub struct Unsafe<O: ?Sized, F: ?Sized, const NAME_HASH: u128> {
+    _marker: PhantomData<O>,
+    // INVARIANT: `field` is only modified via public `unsafe` methods. User code is never
+    // invoked implicitly except via public `unsafe` methods.
+    field: F,
+}
+
+// NOTE on design: It may seem counter-intuitive to offer an impl of traits that
+// don't require `unsafe` to call. Unfortunately, this is a fundamental
+// requirement if users want to be able to mark their types as `Copy`. Luckily,
+// we can implement `Copy` (and its unavoidable super-trait, `Clone`) without
+// invoking user code or opening up the possibility of modifying the field. We
+// do this by only implementing `Copy` and `Clone` when `F: Copy`. For `Clone`,
+// the user is still able to provide a manual impl, so this does not
+// fundamentally restrict what behavior can be supported.
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Copy for Unsafe<O, F, { NAME_HASH }> {}
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Clone for Unsafe<O, F, { NAME_HASH }> {
+    #[allow(clippy::non_canonical_clone_impl)]
+    fn clone(&self) -> Self {
+        // SAFETY: We don't call any user-defined code here (only make a
+        // bit-for-bit copy of `self.field`), so there's no way to accidentally
+        // invoke user-defined code or modify `self.field`.
+        Unsafe { _marker: PhantomData, field: self.field }
+    }
+}
+
+impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
+    /// Gets a reference to the inner value.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn as_ref(&self) -> &F {
+        // SAFETY: This method is unsafe to call.
+        &self.field
+    }
```

Comment:
With `Immutable`, we could provide `as_ref` safely. Perhaps — in the interest of leaving that name open for the safe accessor — this should be called something like `as_ref_unchecked`?

Reply:
IMO we should keep this totally separate from zerocopy. We can also easily release breaking changes since this will never be in anyone's API, so I don't think we need to be too concerned about predicting future API changes.

Reply:
Couldn't zerocopy be an option dependency of this?